applications/RomApplication/python_scripts/rom_analysis.py

```diff
@@ -46,78 +86,67 @@ def ModifyAfterSolverInitialize(self):
-                    self.ResidualUtilityObject = KratosROM.RomResidualsUtility(
-                        computing_model_part,
-                        self.project_parameters["solver_settings"]["rom_settings"],
-                        self._GetSolver()._GetScheme())
+            nodal_modes = self.rom_parameters["nodal_modes"]
+            nodal_dofs = len(self.project_parameters["solver_settings"]["rom_settings"]["nodal_unknowns"].GetStringArray())
+            rom_dofs = self.project_parameters["solver_settings"]["rom_settings"]["number_of_rom_dofs"].GetInt()
+
+            # Set the nodal ROM basis
+            aux = KratosMultiphysics.Matrix(nodal_dofs, rom_dofs)
+            for node in computing_model_part.Nodes:
+                node_id = str(node.Id)
+                for j in range(nodal_dofs):
+                    for i in range(rom_dofs):
+                        aux[j,i] = nodal_modes[node_id][j][i].GetDouble()
+                node.SetValue(KratosROM.ROM_BASIS, aux)
+
+            # Check for HROM stages
+            if self.train_hrom:
+                # Create the training utility to calculate the HROM weights
+                self.__hrom_training_utility = HRomTrainingUtility(
+                    self._GetSolver(),
+                    self.rom_parameters)
+            elif self.run_hrom:
+                # Set the HROM weights in elements and conditions
+                hrom_weights_elements = self.rom_parameters["elements_and_weights"]["Elements"]
+                for key,value in zip(hrom_weights_elements.keys(), hrom_weights_elements.values()):
+                    computing_model_part.GetElement(int(key)+1).SetValue(KratosROM.HROM_WEIGHT, value.GetDouble()) #FIXME: FIX THE +1
+
+                hrom_weights_condtions = self.rom_parameters["elements_and_weights"]["Conditions"]
+                for key,value in zip(hrom_weights_condtions.keys(), hrom_weights_condtions.values()):
+                    computing_model_part.GetCondition(int(key)+1).SetValue(KratosROM.HROM_WEIGHT, value.GetDouble()) #FIXME: FIX THE +1
 
         def FinalizeSolutionStep(self):
-            if self.hyper_reduction_element_selector:
-                if self.hyper_reduction_element_selector.Name == "EmpiricalCubature":
-                    KratosMultiphysics.Logger.PrintInfo("RomAnalysis","Generating matrix of residuals.")
-                    res_mat = self.ResidualUtilityObject.GetResiduals()
-                    np_res_mat = np.array(res_mat, copy=False)
-                    self.time_step_residual_matrix_container.append(np_res_mat)
-
+            # Call the HROM training utility to append the current step residuals
+            # Note that this needs to be done prior to the other processes to avoid unfixing the BCs
+            if self.train_hrom:
+                self.__hrom_training_utility.AppendCurrentStepResiduals()
+
+            # #FIXME: Make this optional. This must be a process
```

Comment:
I'm confused with this, isn't it always projected in ProjectToFineBasis?

Reply:
The point is that the visualization model parts must be independent to the HROM one. I'll take care of this in an upcoming PR. So far I only included some tools required for the creation of such model parts.